File: lua/tst/window-clip.lua
-- A clip set on the WINDOW (parent) layer must scissor the world -> window
-- composite, not just direct draws. Currently the composite ignores it, so
-- the world overwrites the whole window (full red instead of a red box).
--
-- Two blocks pin the same behaviour in both modes:
--   01 PLAIN  -- green comes from the window effect (clear) colour, since a
--                plain present clears the window every frame.
--   02 EXPERT -- green is drawn onto the window layer and persists, since an
--                expert present does not clear the window.

require 'pico.check'

pico.init(true)
pico.set.window { title="Window Clip" }
pico.set.pencil { style='fill' }

local full = {'%', x=0.5, y=0.5, w=1.0, h=1.0, anchor='C'}
local box  = {'%', x=0.5, y=0.5, w=0.5, h=0.5, anchor='C'}

--------------------------------------------------------------------------------
-- PLAIN: green backdrop via the window effect (clear) colour
--------------------------------------------------------------------------------

-- GREEN backdrop: the plain present clears the window to this each frame
pico.set.layer("window")
pico.set.effect { color='green' }

-- clip the WINDOW to a centred box; the composite should honour it
pico.set.scene { clip=box }

-- WORLD: full red -> should reach the window only inside the box
pico.set.layer("world")
pico.set.pencil { color='red' }
pico.output.draw.rect(full)

pico.check("window-clip-01")
-- want: GREEN field, RED centred box
-- now:  full RED (window clip ignored by the composite)

--------------------------------------------------------------------------------
-- EXPERT: green backdrop drawn onto the window layer (persists)
--------------------------------------------------------------------------------

pico.set.expert(true, 0)

-- GREEN background painted on the WINDOW layer (full)
pico.set.layer("window")
pico.set.scene { clip=full }
pico.set.pencil { color='green' }
pico.output.draw.rect(full)

-- clip the WINDOW to a centred box; the composite should honour it
pico.set.scene { clip=box }

-- WORLD: full red -> should reach the window only inside the box
pico.set.layer("world")
pico.output.clear()
pico.set.pencil { color='red' }
pico.output.draw.rect(full)

pico.output.present()
pico.check("window-clip-02")
-- want: GREEN field, RED centred box
-- now:  full RED (window clip ignored by the composite)

pico.init(false)

File: tst/window-clip.c
// A clip set on the WINDOW (parent) layer must scissor the world -> window
// composite, not just direct draws. Currently the composite ignores it, so
// the world overwrites the whole window (full red instead of a red box).
//
// Two blocks pin the same behaviour in both modes:
//   01 PLAIN  -- green comes from the window effect (clear) colour, since a
//                plain present clears the window every frame.
//   02 EXPERT -- green is drawn onto the window layer and persists, since an
//                expert present does not clear the window.
//
// Build/run:  make T=window-clip gen   (inspect tst/out/window-clip-*.png)

#include "pico.h"
#include "../check.h"

int main (void) {
    Pico_Rel_Rect full = { '%', {0.5,0.5, 1.0,1.0}, PICO_ANCHOR_C };
    Pico_Rel_Rect box  = { '%', {0.5,0.5, 0.5,0.5}, PICO_ANCHOR_C };
    Pico_Color green = { 0x00,0xFF,0x00,0xFF };
    Pico_Color red   = { 0xFF,0x00,0x00,0xFF };

    pico_init(1);
    pico_set_window_title("Window Clip");
    pico_set_pencil_style(PICO_STYLE_FILL);

    ///////////////////////////////////////////////////////////////////////////
    // PLAIN: green backdrop via the window effect (clear) colour
    ///////////////////////////////////////////////////////////////////////////

    // GREEN backdrop: the plain present clears the window to this each frame
    pico_set_layer("window");
    pico_set_effect_color(green);

    // clip the WINDOW to a centred box; the composite should honour it
    pico_set_scene_clip(box);

    // WORLD: full red -> should reach the window only inside the box
    pico_set_layer("world");
    pico_set_pencil_color(red);
    pico_output_draw_rect(full);

    _pico_check("window-clip-01");
    // want: GREEN field, RED centred box
    // now:  full RED (window clip ignored by the composite)

    ///////////////////////////////////////////////////////////////////////////
    // EXPERT: green backdrop drawn onto the window layer (persists)
    ///////////////////////////////////////////////////////////////////////////

    pico_set_expert(1, 0);

    // GREEN background painted on the WINDOW layer (full)
    pico_set_layer("window");
    pico_set_scene_clip(full);
    pico_set_pencil_color(green);
    pico_output_draw_rect(full);

    // clip the WINDOW to a centred box; the composite should honour it
    pico_set_scene_clip(box);

    // WORLD: full red -> should reach the window only inside the box
    pico_set_layer("world");
    pico_output_clear();
    pico_set_pencil_color(red);
    pico_output_draw_rect(full);

    pico_output_present(1);
    _pico_check("window-clip-02");
    // want: GREEN field, RED centred box
    // now:  full RED (window clip ignored by the composite)

    pico_init(0);
    return 0;
}
